src/modules/fancyzones/lib/ZoneWindow.cpp

```diff
@@ -407,30 +426,69 @@ IFACEMETHODIMP ZoneWindow::MoveSizeEnd(HWND window, POINT const& ptScreen) noexc
 {
     RestoreOrginalTransparency();
 
-    if (m_windowMoveSize != window)
+    std::shared_lock lock(m_mutex);
+    auto windowMoveSize = m_windowMoveSize;
+    auto thisWindow = m_window.get();
+    auto activeZoneSet = m_activeZoneSet;
+    lock.unlock();
+
+    if (windowMoveSize != window)
     {
         return E_INVALIDARG;
     }
 
-    if (m_activeZoneSet)
+    if (activeZoneSet)
     {
         POINT ptClient = ptScreen;
-        MapWindowPoints(nullptr, m_window.get(), &ptClient, 1);
-        m_activeZoneSet->MoveWindowIntoZoneByPoint(window, m_window.get(), ptClient);
+        MapWindowPoints(nullptr, thisWindow, &ptClient, 1);
+        activeZoneSet->MoveWindowIntoZoneByPoint(window, thisWindow, ptClient);
 
         SaveWindowProcessToZoneIndex(window);
     }
-    Trace::ZoneWindow::MoveSizeEnd(m_activeZoneSet);
+    Trace::ZoneWindow::MoveSizeEnd(activeZoneSet);
 
     HideZoneWindow();
+    std::unique_lock writeLock(m_mutex);
     m_windowMoveSize = nullptr;
     return S_OK;
 }
 
+IFACEMETHODIMP_(bool)
+ZoneWindow::IsDragEnabled() noexcept
+{
+    std::shared_lock lock(m_mutex);
+    return m_dragEnabled;
+}
+
+IFACEMETHODIMP_(std::wstring)
+ZoneWindow::UniqueId() noexcept
+{
+    std::shared_lock lock(m_mutex);
+    return m_uniqueId;
+}
+
+IFACEMETHODIMP_(std::wstring)
+ZoneWindow::WorkAreaKey() noexcept
+{
+    std::shared_lock lock(m_mutex);
+    return m_workArea;
+}
+
+IFACEMETHODIMP_(IZoneSet*)
+ZoneWindow::ActiveZoneSet() noexcept
+{
+    std::shared_lock lock(m_mutex);
+    return m_activeZoneSet.get();
+}
+
 IFACEMETHODIMP_(void)
 ZoneWindow::RestoreOrginalTransparency() noexcept
 {
-    if (m_host->isMakeDraggedWindowTransparentActive() && draggedWindow != nullptr)
+    std::shared_lock lock(m_mutex);
+    auto hostTransparentActive = m_host->isMakeDraggedWindowTransparentActive();
+    lock.unlock();
+
+    if (hostTransparentActive && draggedWindow != nullptr)
     {
         SetLayeredWindowAttributes(draggedWindow, draggedWindowCrKey, draggedWindowInitialAlpha, draggedWindowDwFlags);
         SetWindowLong(draggedWindow, GWL_EXSTYLE, draggedWindowExstyle);

```

Comment:
Did you leave this members (draggedWindow, draggedWindowCrKey, ...) outside of mutex range on puprose?

Reply:
No, I missed the fact they are members because they didn't have the `m_` prefix. I'll rename them and make sure they are safe.